mm2src/coins/z_coin/storage/walletdb/mod.rs

```diff
@@ -33,27 +46,115 @@ pub struct WalletDbShared {
+        };
+
+        let sync_block = match zcoin_builder.z_coin_params.mode.clone() {
+            ZcoinRpcMode::Light { sync_params, .. } => {
+                let sync_height = match sync_params {
+                    Some(params) => {
+                        if let Some(date) = params.date {
+                            let current_block_height = rpc
+                                .get_block_height()
+                                .await
+                                .map_err(|err| WalletDbError::GrpcError(err.to_string()))?;
+
+                            let starting_height = calculate_starting_height_from_date(date, current_block_height)
+                                .map_err(|err| {
+                                    WalletDbError::ZcoinClientInitError(ZcoinClientInitError::ZcashDBError(err))
+                                })?;
+
+                            info!(
+                        "Found date in sync params for {}. Walletdb will be built using block height {} as the \
+                        starting \
+                        block for scanning and updating walletdb",
+                        zcoin_builder.ticker, starting_height
+                    );
+                            Some(starting_height)
+                        } else {
+                            if params.height.is_some() {
+                                info!(
+                            "Found height in sync params for {}. Walletdb will be built using block height {:?} as the \
+                            starting block for scanning and updating walletdb",
+                            zcoin_builder.ticker, params.height
+                        );
+                            }
+                            params.height
+                        }
+                    },
+                    None => None,
+                };
+
+                match sync_height {
+                    Some(height) => {
+                        let block = rpc
+                            .get_tree_state(height)
+                            .await
+                            .map_err(|err| WalletDbError::GrpcError(err.to_string()))?;
+                        let hash = H256::from_str(&block.hash)
+                            .map_err(|err| WalletDbError::DecodeError(err.to_string()))?
+                            .reversed();
+                        let sapling_tree = Bytes::new(
+                            FromHex::from_hex(&block.tree)
+                                .map_err(|err: FromHexError| WalletDbError::DecodeError(err.to_string()))?,
+                        );
+
+                        Some(CheckPointBlockInfo {
+                            height: block.height as u32,
+                            hash,
+                            time: block.time,
+                            sapling_tree,
+                        })
+                    },
+                    None => zcoin_builder.protocol_info.check_point_block.clone(),
```

Comment:
CA only suggested this and using default(current_height)

Reply:
Either of the two options were suggested so let's leave it with `current block height` for now. I think it's trivial to change this if `block height minus some buffer` was needed instead.

Reply:
> if protocol_info.check_point_block is None

What I meant by this is that instead of `zcoin_builder.protocol_info.check_point_block.clone(),` it should be
```rust
None => match zcoin_builder.protocol_info.check_point_block {
    Some(ref check_point_block) => Some(check_point_block.clone()),
    None => {
        // This block of code is repeated it should be refactored into a function check_point_block_from_height
        let block = rpc
            .get_tree_state(current_block_height)
            .await
            .map_err(|err| WalletDbError::GrpcError(err.to_string()))?;
        let hash = H256::from_str(&block.hash)
            .map_err(|err| WalletDbError::DecodeError(err.to_string()))?
            .reversed();
        let sapling_tree = Bytes::new(
            FromHex::from_hex(&block.tree)
                .map_err(|err: FromHexError| WalletDbError::DecodeError(err.to_string()))?,
        );

        Some(CheckPointBlockInfo {
            height: block.height as u32,
            hash,
            time: block.time,
            sapling_tree,
        })
    },
},
```